src/charm.py

```diff
@@ -111,274 +69,92 @@ def __init__(self, *args):
 
-    def _on_secret_changed(self, event: SecretChangedEvent) -> None:
-        """Handler for `secret_changed` events."""
-        if not event.secret.label or not self.state.cluster.relation:
-            return
+    def _set_status(self, key: Status) -> None:
+        """Sets charm status."""
+        status: StatusBase = key.value.status
+        log_level: DebugLevel = key.value.log_level
+
+        getattr(logger, log_level.lower())(status.message)
+        self.unit.status = status
+
+    def _on_config_changed(self, _):
+        """Handler for `config_changed` events.
 
-        if event.secret.label == self.state.cluster.data_interface._generate_secret_label(
-            PEER,
-            self.state.cluster.relation.id,
-            "extra",  # pyright: ignore[reportArgumentType] -- Changes with the https://github.com/canonical/data-platform-libs/issues/124
+        This handler is in charge of stopping the workloads, since the sub-operators would not
+        be instantiated if roles are changed.
+        """
+        if BROKER.value not in self.config.roles and self.broker.workload.active():
+            self.broker.workload.stop()
+
+        if (
+            BALANCER.value not in self.config.roles
+            and self.unit.is_leader()
+            and self.broker.workload.active()
         ):
-            self.on.config_changed.emit()
-
-    def _on_storage_attached(self, event: StorageAttachedEvent) -> None:
-        """Handler for `storage_attached` events."""
-        # new dirs won't be used until topic partitions are assigned to it
-        # either automatically for new topics, or manually for existing
-        # set status only for running services, not on startup
-        self.workload.exec(f"chmod -R 750 {self.workload.paths.data_path}")
-        self.workload.exec(f"chown -R {USER}:{GROUP} {self.workload.paths.data_path}")
-        self.workload.exec(
-            f"""find {self.workload.paths.data_path} -type f -name "meta.properties" -delete || true"""
-        )
-        if self.workload.active():
-            self._set_status(Status.ADDED_STORAGE)
-            self._on_config_changed(event)
-
-    def _on_storage_detaching(self, _: StorageDetachingEvent) -> None:
-        """Handler for `storage_detaching` events."""
-        # in the case where there may be replication recovery may be possible
-        if self.state.brokers and len(self.state.brokers) > 1:
-            self._set_status(Status.REMOVED_STORAGE)
-        else:
-            self._set_status(Status.REMOVED_STORAGE_NO_REPL)
+            self.balancer.workload.stop()
 
-        self.on.config_changed.emit()
+    def _restart_broker(self, event: EventBase) -> None:
+        """Handler for `rolling_ops` restart events.
 
-    def _restart(self, event: EventBase) -> None:
-        """Handler for `rolling_ops` restart events."""
+        The RollingOpsManager expecting a charm instance, we cannot move this method to the broker logic.
+        """
```

Comment:
If I understand your question correctly, yes, we could, but we would need to have `_disable_enable_restart_broker` in charm.py anyway because `callback_override` is a str parameter. So I figured we might as well keep the restart logic in this file